### PR TITLE
feat: support regex in path

### PR DESCRIPTION
**Pull request by lxm**
Please answer these questions before submitting a pull request

- Why submit this pull request?
- [x] New feature provided

- Related issues #768 

___
### New feature or improvement
- support a new annotation `k8s.apisix.apache.org/use-regex` which to change path behavior using apisix's vars

### CODE REVIEW

File: pkg/kube/translation/annotations/types.go [hunk 1]
```diff
@@ -18,6 +18,10 @@ import (
 	"strings"
 )
 
+const (
+	AnnotationsPrefix = "k8s.apisix.apache.org/"
```
Comment (tokers) [inline pkg/kube/translation/annotations/types.go:22]: Is this constant exported to the outside? If not, we can rename it to `annotationsPrefix`.

File: pkg/kube/translation/ingress.go [hunk 2]
```diff
@@ -87,29 +90,52 @@ func (t *translator) translateIngressV1(ing *networkingv1.Ingress) (*TranslateCo
...
+					uris = append(uris, prefix)
+				} else if *pathRule.PathType == networkingv1.PathTypeExact {
+					uris = append(uris, pathRule.Path)
+				} else if *pathRule.PathType == networkingv1.PathTypeImplementationSpecific && useRegex {
+					nginxVars = append(nginxVars, configv2alpha1.ApisixRouteHTTPMatchExpr{
+						Subject: configv2alpha1.ApisixRouteHTTPMatchExprSubject{
+							Scope: configv2alpha1.ScopePath,
+						},
+						Op:    configv2alpha1.OpRegexMatch,
+						Set:   []string{},
```
Comment (tokers) [inline pkg/kube/translation/ingress.go:120]: We don't initialize this field.

File: pkg/kube/translation/ingress.go [hunk 3]
```diff
@@ -175,29 +203,52 @@ func (t *translator) translateIngressV1beta1(ing *networkingv1beta1.Ingress) (*T
...
+					uris = append(uris, prefix)
+				} else if *pathRule.PathType == networkingv1beta1.PathTypeExact {
+					uris = append(uris, pathRule.Path)
+				} else if *pathRule.PathType == networkingv1beta1.PathTypeImplementationSpecific && useRegex {
+					nginxVars = append(nginxVars, configv2alpha1.ApisixRouteHTTPMatchExpr{
+						Subject: configv2alpha1.ApisixRouteHTTPMatchExprSubject{
+							Scope: configv2alpha1.ScopePath,
+						},
+						Op:    configv2alpha1.OpRegexMatch,
+						Set:   []string{},
```
Comment (tokers) [inline pkg/kube/translation/ingress.go:233]: Ditto.

File: pkg/kube/translation/ingress.go [hunk 4]
```diff
@@ -266,29 +319,52 @@ func (t *translator) translateIngressExtensionsV1beta1(ing *extensionsv1beta1.In
...
+					uris = append(uris, prefix)
+				} else if *pathRule.PathType == extensionsv1beta1.PathTypeExact {
+					uris = append(uris, pathRule.Path)
+				} else if *pathRule.PathType == extensionsv1beta1.PathTypeImplementationSpecific && useRegex {
+					nginxVars = append(nginxVars, configv2alpha1.ApisixRouteHTTPMatchExpr{
+						Subject: configv2alpha1.ApisixRouteHTTPMatchExprSubject{
+							Scope: configv2alpha1.ScopePath,
+						},
+						Op:    configv2alpha1.OpRegexMatch,
+						Set:   []string{},
```
Comment (tokers) [inline pkg/kube/translation/ingress.go:349]: Ditto.

**Review by tokers — COMMENTED** [review]
We may also add some test cases.

File: pkg/kube/translation/annotations/types.go (hunk 1 above)
Comment (lxm) [reply pkg/kube/translation/annotations/types.go:22]: this will be used in `translation` package to dicide whether use-regex annotation is true

**Comment by lxm** [comment]
> We may also add some test cases.

have written more test cases.
retested the feature

File: pkg/kube/translation/annotations/types.go (hunk 1 above)
Comment (tokers) [reply pkg/kube/translation/annotations/types.go:22]: OK, let's add some comments for exportable constants.

**Comment by tokers** [comment]
> > We may also add some test cases.
> 
> have written more test cases. retested the feature

By say adding test cases, we'd better also add the E2E test cases.

**Comment by lxm** [comment]
> > > We may also add some test cases.
> > 
> > 
> > have written more test cases. retested the feature
> 
> By say adding test cases, we'd better also add the E2E test cases.

working on it.

File: pkg/kube/translation/annotations/types.go (hunk 1 above)
Comment (lxm) [reply pkg/kube/translation/annotations/types.go:23]: done

**Comment by lxm** [comment]
@tokers @tao12345666333 @lingsamuel @gxthrj  

finally got the e2e testing work.
now, this PR needs a review again.

**Comment by tao12345666333** [comment]
In addition, can you please add documentation to introduce this feature?

**Comment by lxm** [comment]
> In addition, can you please add documentation to introduce this feature?

sure.
Because `exprs` is complex, more configurations may be added in the future.

File: pkg/kube/translation/ingress.go [hunk 5]
```diff
@@ -87,29 +94,49 @@ func (t *translator) translateIngressV1(ing *networkingv1.Ingress) (*TranslateCo
...
+						prefix += "/*"
+					}
+					uris = append(uris, prefix)
+				} else if *pathRule.PathType == networkingv1.PathTypeImplementationSpecific && useRegex {
+					nginxVars = append(nginxVars, configv2alpha1.ApisixRouteHTTPMatchExpr{
+						Subject: configv2alpha1.ApisixRouteHTTPMatchExprSubject{
+							Scope: configv2alpha1.ScopePath,
+						},
+						Op:    configv2alpha1.OpRegexMatch,
+						Value: &pathRule.Path,
```
Comment (tokers) [inline pkg/kube/translation/ingress.go:122]: I think we should validate the regex pattern in the admission server.

Comment (lxm) [reply pkg/kube/translation/ingress.go:122]: this path comes from ingress resources, I don't thinke we need to do this.

Comment (tokers) [reply pkg/kube/translation/ingress.go:122]: Why? Does the Ingress resource validator validate it?

Comment (tokers) [reply pkg/kube/translation/ingress.go:122]: > 

Even it comes from the Ingress, we still have to validate it before translating, or it cannot be used by APISIX.

Comment (lxm) [reply pkg/kube/translation/ingress.go:122]: currently, the apisix-ingress-controller will not validate all the ingress resource being created? Maybe valiate the ingress resource should be another feature.

On the other hand, a issue in kubernetes projects https://github.com/kubernetes/kubernetes/issues/90210
the validation was removed

Comment (tokers) [reply pkg/kube/translation/ingress.go:122]: All right, that'd be unexpected, would you like to submit an issue? I think we can implement it in the future.

Comment (lxm) [reply pkg/kube/translation/ingress.go:122]: ok

File: pkg/kube/translation/ingress.go [hunk 6]
```diff
@@ -29,16 +29,23 @@ import (
 
 	"github.com/apache/apisix-ingress-controller/pkg/id"
 	apisixv12 "github.com/apache/apisix-ingress-controller/pkg/kube/apisix/apis/config/v1"
+	configv2alpha1 "github.com/apache/apisix-ingress-controller/pkg/kube/apisix/apis/config/v2alpha1"
```
Comment (gxthrj) [inline pkg/kube/translation/ingress.go:32]: Please merge master， the `v2alpha1` no longer exist, use `v2beta3` instead.

Comment (lxm) [reply pkg/kube/translation/ingress.go]: a lot of changes were made.
I've done the changes

**Review by tokers — APPROVED** [review]

**Comment by lxm** [comment]
> In addition, can you please add documentation to introduce this feature?

done for docs

**Comment by lxm** [comment]
@gxthrj @tao12345666333 need a review

I think the e2e failure has nothing to do with this pr

**Comment by tao12345666333** [comment]
re-run all jobs.

**Comment by lxm** [comment]
> re-run all jobs.

E2E tests failed with "check the ingress lb status is updated" this time, maybe something wrong with test cases. I can pass this case singly in a private environment.

**Comment by lxm** [comment]
ping @tao12345666333 @gxthrj @lingsamuel

**Review by tao12345666333 — APPROVED** [review]
LGTM

**Comment by tao12345666333** [comment]
Thd code LGTM

I added the label of `do-not-merge/hold` to it because this feature is not planned to be in the v1.4 version.

Do you have a strong demand for this feature? @lxm

**Comment by lxm** [comment]
> Thd code LGTM
> 
> I added the label of `do-not-merge/hold` to it because this feature is not planned to be in the v1.4 version.
> 
> Do you have a strong demand for this feature? @lxm

right, I want to use apisix-ingress-controller to replace the nginx-ingress-controller

without this feature, this can not be done.

**Comment by tao12345666333** [comment]
ok I understand your situation, I have added it to v1.4.

**Pull request merged**